server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1124,48 +1129,59 @@ private WildcardExpressionResolver() {
+            Objects.requireNonNull(expressions);
             // only check open/closed since if we do not expand to open or closed it doesn't make sense to
             // expand to hidden
             if (context.getOptions().expandWildcardsClosed() == false && context.getOptions().expandWildcardsOpen() == false) {
                 return expressions;
+            } else if (isEmptyOrTrivialWildcard(expressions)) {
+                return innerResolveAll(context);
+            } else {
+                return innerResolve(context, expressions);
             }
+        }
 
-            if (isEmptyOrTrivialWildcard(expressions)) {
-                List<String> resolvedExpressions = resolveEmptyOrTrivialWildcard(context);
-                if (context.includeDataStreams()) {
-                    final Map<String, IndexAbstraction> dataStreamsAbstractions = context.getState()
-                        .metadata()
-                        .getIndicesLookup()
-                        .entrySet()
-                        .stream()
-                        .filter(entry -> entry.getValue().getType() == IndexAbstraction.Type.DATA_STREAM)
-                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-                    // dedup backing indices if expand hidden indices option is true
-                    Set<String> resolvedIncludingDataStreams = new HashSet<>(resolvedExpressions);
-                    resolvedIncludingDataStreams.addAll(
-                        expand(context, dataStreamsAbstractions, expressions.isEmpty() ? "_all" : expressions.get(0))
-                    );
-                    return new ArrayList<>(resolvedIncludingDataStreams);
-                }
+        /**
+         * Returns all the indices and all the datastreams, considering the open/closed, system, and hidden context parameters.
+         * Depending on the context, returns the names of the datastreams themselves or their backing indices.
+         */
+        private static Collection<String> innerResolveAll(Context context) {
+            List<String> resolvedExpressions = resolveEmptyOrTrivialWildcard(context);
+            if (context.includeDataStreams() == false) {
                 return resolvedExpressions;
+            } else {
+                final List<IndexAbstraction> dataStreamsAbstractions = context.getState()
+                    .metadata()
+                    .getIndicesLookup()
+                    .values()
+                    .stream()
+                    .filter(indexAbstraction -> indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM)
+                    .collect(Collectors.toList());
+                // dedup backing indices if expand hidden indices option is true
+                Set<String> resolvedIncludingDataStreams = new HashSet<>(resolvedExpressions);
+                expandMatches(context, dataStreamsAbstractions, false, resolvedIncludingDataStreams::add);
```

Comment:
Side note, I think the `expandMatches` here should eventually be removed, but that's a conversation to have another time.

Wildcard expressions that start with "." must cover hidden indices that start with "." . In this case, the expression is "*" or "_all" so it's not a wildcard that starts with "." so it will not cover hidden indices that with ".", which is the meaning of the `false` flag param.